bridge/src/reactor.rs

```diff
@@ -0,0 +1,124 @@
+
+#[derive(Debug)]
+pub enum ReactorError {
+    ConnectionError(ConnectionError)
+}
+
+impl std::fmt::Display for ReactorError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ReactorError::ConnectionError(error) => write!(f, "{}", error),
+        }
+    }
+}
+
+impl From<ConnectionError> for ReactorError {
+    fn from(err: ConnectionError) -> Self {
+        ReactorError::ConnectionError(err)
+    }
+}
+
+impl std::error::Error for ReactorError {}
+
+pub type ReactorResult = Result<Response, ReactorError>;
+
+// A request wrapper which includes a channel for sending back the response
+#[derive(Debug)]
+struct Envelope {
+    request: Request,
+    channel: oneshot::Sender<ReactorResult>
+}
+
+async fn handle_request(request: Request, channel: oneshot::Sender<ReactorResult>) {
+    match request {
+        Request::Connect { host } => {
+            let result = Connection::connect(host).await
+                .map(|connection| { Response::Connection { connection: connection } })
+                .map_err(|error| ReactorError::ConnectionError(error));
+
+            channel.send(result);
+        },
+        Request::Rpc { mut connection, method, bytes } => {
+            let result = connection.call(&method, bytes).await
+                .map(|bytes| { Response::Rpc { bytes: bytes } })
+                .map_err(|error| ReactorError::ConnectionError(error));
+
+            channel.send(result);
+        }
+    }
+}
+
+fn start_loop(thread_count: usize, receiver: &mut mpsc::UnboundedReceiver<Envelope>) {
+    let runtime = Builder::new_multi_thread()
+        .worker_threads(thread_count)
+        .enable_all()
+        .thread_name("core")
+        .build()
+        .unwrap();
+
+    runtime.block_on(async move {
+        while let Some(msg) = receiver.recv().await {
+            tokio::spawn(handle_request(msg.request, msg.channel));
+        }
+    });
+}
+
+// A Reactor is responsible for processing requests asynchronously using an event loop
+// while exposing a synchronous interface to the callers
+
+#[derive(Clone)]
+pub struct Reactor {
+    sender: mpsc::UnboundedSender<Envelope>,
+}
+
+impl Reactor {
+    pub fn new(thread_count: usize) -> Reactor {
+        let (tx, mut rx) = mpsc::unbounded_channel::<Envelope>();
+
+        std::thread::spawn(move || start_loop(thread_count, &mut rx));
+
+        Reactor { sender: tx }
+    }
+
+    pub fn process(&self, request: Request) -> ReactorResult {
+        // Create a one-off channel to communicate the result back
+        let (tx, rx) = oneshot::channel();
+        let envelope = Envelope { request: request, channel: tx };
+
+        match self.sender.send(envelope) {
+            Ok(()) => println!("task sent"),
+            Err(_) => panic!("The shared runtime has shut down."),
+        }
+
+        // Block until a result is sent back (the timing is determined by the task handler)
+        let response = rx.blocking_recv();
```

Comment:
Not sure this will work. It means the calls here https://github.com/temporalio/sdk-ruby/pull/22/files#diff-8168c62c02ae142c498a22a84d62c19aef95449311bd0e6f60b9646c4eb5c798R45 will block, which means any long running call will tie up the caller.

If the way Ruby is calling into these means it won't tie up all other execution while calling into the native code, then that's fine and we can keep doing this - however for that to be true it'd mean we would need every native call to run on a dedicated thread to avoid blocking ongoing work, and that also seems wrong.

So I suspect this needs to change. Rather than blocking on this channel here, we will need some way for the bridge methods to pass something that can be fulfilled with the result via a callback or something, and that can happen inside of the async task instead of it sending down this channel.

We can zoom for a bit to discuss if you like, lemme know

Reply:
> Rather than blocking on this channel here, we will need some way for the bridge methods to pass something that can be fulfilled with the result via a callback or something, and that can happen inside of the async task instead of it sending down this channel.

:+1: In Ruby I think the rule of thumb is "block a thread, that's fine, if they want async they can use another approach" (at least for now until fibers), which means we might as well block on tokio executions right from the calls instead of using the channel.